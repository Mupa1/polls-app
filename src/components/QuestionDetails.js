import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import ProgressBar from '@ramonak/react-progress-bar';

import { getQuestionDetails } from '../api';

const QuestionDetails = (props) => {
  const dispatch = useDispatch();
  const selectedQuestion = useSelector((state) => state.question.question);
  const { question, choices } = selectedQuestion;

  useEffect(() => {
    const { questionId } = props.match.params;

    dispatch(getQuestionDetails(questionId));
  }, []);

  const handleVotesPercentage = (choices, votes) => {
    const initialValue = 0;
    const sum = choices.reduce(
      (accumulator, currentValue) => accumulator + currentValue.votes,
      initialValue,
    );

    const percentage = ((votes / sum) * 100).toFixed(2);
    return percentage;
  };

  return (
    <section className="details">
      {selectedQuestion && (
        <>
          <h2>
            Question:
            {' '}
            {question}
          </h2>

          {choices.map((choice) => (
            <div key={choice.url} className="choices-card">
              <div>{choice.choice}</div>
              <div>
                Votes:
                {' '}
                {choice.votes}
              </div>
              <div>
                {handleVotesPercentage(choices, choice.votes)}
                {' '}
                %
              </div>
              <div>
                <ProgressBar
                  completed={handleVotesPercentage(choices, choice.votes)}
                  isLabelVisible={false}
                />
              </div>
            </div>
          ))}
        </>
      )}
    </section>
  );
};

QuestionDetails.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      questionId: PropTypes.string,
    }),
  }).isRequired,
};

export default QuestionDetails;
