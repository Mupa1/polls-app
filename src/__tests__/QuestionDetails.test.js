/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import store from '../redux/store';
import QuestionDetails from '../components/QuestionDetails';

const props = {
  match: {
    params: {
      questionId: '34',
    },
  },
  history: {
    push: () => '/',
  },
};

describe('QuestionDetails component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(
      <Provider store={store}>
        <QuestionDetails {...props} />
      </Provider>, div,
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <QuestionDetails {...props} />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
