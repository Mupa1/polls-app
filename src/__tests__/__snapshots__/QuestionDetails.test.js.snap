// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionDetails component matches snapshot 1`] = `
<section
  className="details"
>
  
</section>
`;
